pkg/proxy/proxy.go

```diff
@@ -0,0 +1,319 @@
+		code, 0, l.timeDiff.Seconds())
+}
+
+func (p *Proxy) CreateOrUpdateRedirect(l4 *policy.L4Filter, id string, source ProxySource) (*Redirect, error) {
+	fwd, err := forward.New()
+	if err != nil {
+		return nil, err
+	}
+
+	if strings.ToLower(l4.Redirect) != "http" {
+		return nil, fmt.Errorf("unknown L7 protocol \"%s\"", l4.Redirect)
+	}
+
+	for _, r := range l4.Rules {
+		if !route.IsValid(r.Expr) {
+			return nil, fmt.Errorf("invalid filter expression: %s", r.Expr)
+		}
+	}
+
+	gcOnce.Do(func() {
+		if lf := viper.GetString("access-log"); lf != "" {
+			if logFile, err = os.Create(lf); err != nil {
+				log.Warningf("cannot open access log: %s", err)
+			} else {
+				logBuf = bufio.NewWriter(logFile)
+			}
+		}
+
+		go func() {
+			for {
+				time.Sleep(time.Duration(10) * time.Second)
+				if deleted := GC(); deleted > 0 {
+					log.Debugf("Evicted %d entries from proxy table\n", deleted)
+				}
+			}
+		}()
+	})
+
+	p.mutex.Lock()
+
+	if r, ok := p.redirects[id]; ok {
+		r.updateRules(l4.Rules)
+		log.Debugf("updated existing proxy instance %+v", r)
+		p.mutex.Unlock()
+		return r, nil
+	}
+
+	to, err := p.allocatePort()
+	if err != nil {
+		p.mutex.Unlock()
+		return nil, err
+	}
+
+	redir := &Redirect{
+		id:       id,
+		FromPort: uint16(l4.Port),
+		ToPort:   to,
+		source:   source,
+		router:   route.New(),
+	}
+
+	redirect := http.HandlerFunc(func(w http.ResponseWriter, req *http.Request) {
+		startDelta := time.Now()
+		record := &LogRecord{
+			req:       req,
+			timeStart: time.Time{},
+		}
+
+		reason := "no rules"
+
+		// Validate access to L4/L7 resource
+		p.mutex.Lock()
+		if len(redir.Rules) > 0 {
+			rule, _ := redir.router.Route(req)
+			if rule == nil {
+				http.Error(w, "Access denied", http.StatusForbidden)
+				p.mutex.Unlock()
+				redir.Log(record, http.StatusForbidden, "access denied")
+				return
+			} else {
+				ar := rule.(policy.AuxRule)
+				log.Debugf("Allowing request based on rule %+v\n", ar)
+				reason = fmt.Sprintf("rule: %+v", ar)
+			}
+		}
+		p.mutex.Unlock()
+
+		// Reconstruct original URL used for the request
+		if newURL, err := generateURL(w, req, to); err != nil {
+			log.Errorf("%s\n", err)
+			http.Error(w, err.Error(), http.StatusBadRequest)
+			redir.Log(record, http.StatusBadRequest, fmt.Sprintf("cannot generate url: %s", err))
+			return
+		} else {
+			req.URL = newURL
+		}
+
+		fwd.ServeHTTP(w, req)
+		record.timeDiff = time.Now().UTC().Sub(startDelta)
+		redir.Log(record, http.StatusOK, reason)
+	})
+
+	redir.server = manners.NewWithServer(&http.Server{
+		Addr:    fmt.Sprintf(":%d", to),
+		Handler: redirect,
+	})
+
+	redir.updateRules(l4.Rules)
+	p.allocatedPorts[to] = redir
+	p.redirects[id] = redir
+
+	p.mutex.Unlock()
+
+	log.Debugf("Created new proxy intance %+v", redir)
+
+	go func() {
+		redir.server.ListenAndServe()
+	}()
+
+	return redir, nil
+}
+
+func (p *Proxy) GetRedirect(id string) *Redirect {
+	p.mutex.RLock()
+	defer p.mutex.RUnlock()
+
+	return p.redirects[id]
```

Comment:
If you return this and the caller changes the `Redirect` it would be outside the mutex, can you return a copy instead?